Log fatal errors to stderr, add exceptions to 'marked' errors

raboof opened the pull request:
Refs #23107
After logging to stderr it seems reasonable to include the exception
in the logging though the logging subsystem

raboof committed:
Log fatal errors to stderr, add exceptions to 'marked' errors (#23107)

After logging to stderr it seems reasonable to include the exception
in the logging though the logging subsystem

jrudolph commented on akka-actor/src/main/scala/akka/actor/ActorSystem.scala:
```diff
@@ -666,24 +666,34 @@ private[akka] class ActorSystemImpl(
+      private def logFatalError(message: String, cause: Throwable, thread: Thread): Unit = {
+        // First log to stderr as this has the best chance to get through in an 'emergency panic' situation:
+        import System.err
+        err.print("Uncaught error from thread [")
```
Maybe add `cause.getMessage` as well here.

jrudolph commented on akka-actor/src/main/scala/akka/actor/ActorSystem.scala:
```diff
@@ -666,24 +666,34 @@ private[akka] class ActorSystemImpl(
             } else {
-              markerLogging.error(LogMarker.Security, cause, "Uncaught fatal error from thread [{}] shutting down ActorSystem [{}]", thread.getName, name)
+              logFatalError("shutting down", cause, thread)
               terminate()
```
Run `terminate` in the finally block as well?

raboof replied:
`logFatalError` is pretty simple, if that crashes I'm not sure there's much of a chance `terminate()` will do anything useful... might make the code a bit more symmetrical though :)

patriknw replied:
+1 to `finally`

jrudolph commented on akka-actor-tests/src/test/scala/akka/event/MarkerLoggingSpec.scala:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    "allow cause exceptions in error messages" in {
+      class MyException(message: String) extends Exception(message)
```
Woops, great find. This fix could be in an extra commit.

jrudolph reviewed (approved):
Mostly LGTM, small nitpicks.

raboof committed:
Add cause message to logging on fatal error

raboof committed:
Flush twice for extra safety

printStackTrace does some allocations, there might be rare cases where
we can at least salvage the cause.

patriknw commented on akka-actor/src/main/scala/akka/event/Logging.scala:
```diff
@@ -1456,31 +1456,31 @@ class MarkerLoggingAdapter(
    */
   def error(marker: LogMarker, cause: Throwable, template: String, arg1: Any, arg2: Any): Unit =
-    if (isErrorEnabled) bus.publish(Error(logSource, logClass, format(template, arg1, arg2), mdc, marker))
+    if (isErrorEnabled) bus.publish(Error(cause, logSource, logClass, format(template, arg1, arg2), mdc, marker))
```
wow!

patriknw reviewed (approved):
LGTM

raboof committed:
Terminate in a finally block

jrudolph commented on akka-actor/src/main/scala/akka/actor/ActorSystem.scala:
```diff
@@ -671,8 +671,11 @@ private[akka] class ActorSystemImpl(
             } else {
-              logFatalError("shutting down", cause, thread)
-              terminate()
+              try {
```
Nitpick: no braces needed ;)

raboof replied:
that is actually quite a lot nicer :+1:

raboof committed:
No braces

jrudolph commented:
@johanandren just pointed out that #22324 already tracks the MarkerLoggingAdapter issue.
